src/design/introspection.stanza

```diff
@@ -173,9 +219,287 @@ public defn get-pads-from-port (p2p-map:Tuple<KeyValue<JITXObject, Tuple<JITXObj
+  inside pcb-module:
+    val comp? = get-component-from-port(pt, cxt)
+    match(comp?):
+      (_:None):
+        throw $ ValueError("No Parent Component Found for Port '%_'" % [ref(pt)])
+      (given:One<JITXObject>):
+        val p2p-map = port-to-pads(value(given))
+        get-pads-from-port(p2p-map, pt)
+
+doc: \<DOC>
+Retrieve the parent instance (component or module) for a port.
+
+This function expects you to pass something like `comp.GND` which
+is a `port` on `comp` and then this function will get a reference
+to `comp` as a `JITXObject`.
+
+Be careful of child port on bundle ports - these will give a ref
+to the parent bundle port - not the component.
+
+This function expects to be called from `pcb-module` context.
+
+@param pt Port on a component or module
+@return Reference to a `pcb-module` or `pcb-component` instance.
+<DOC>
+public defn get-parent-from-port (pt:JITXObject, cxt:JITXObject = self) -> Maybe<JITXObject>:
+  inside pcb-module:
+    val pt-R = ref(pt)
+    val parent-R? = get-parent-ref(pt-R)
+    match(parent-R?):
+      (_:None): One(self)
+      (given:One<Ref>):
+        find-instance-by-ref(value(given), cxt)
+
+
+doc: \<DOC>
+Get a component instance from a Port
+
+This function can be called from a `pcb-module` context. Calling from
+a `pcb-component` context will likely result in an exception.
+
+@param pt Port on a component instance.
+@param cxt Context to search for components in.
+@return If we
+<DOC>
+public defn get-component-from-port (pt:JITXObject, cxt:JITXObject = self) -> Maybe<JITXObject>:
+  inside pcb-module:
+    val pt-R = ref(pt)
+    val parent-R? = get-parent-ref(pt-R)
+    match(parent-R?):
+      (_:None):
+        ; TODO - I need a way to differentiate between `pcb-module` and `pcb-component`
+        ;   here otherwise, this component might return a module instance breaking the
+        ;   abstraction.
+        throw $ ValueError("UnHandle-able Case - Did you pass a module port or from a `pcb-component` context?")
+        ; One(self)
+      (given:One<Ref>):
+        val parent-R = value(given)
+        ; We need to search because if the user were to have passed
+        ;  a SinglePin port of a bundle port - then we need to continue looking up the tree
+        ;  until we find the component instance and not just bundle port.
+        val trail-comp? = find-instance-by-ref(parent-R, cxt)
+        ; I want to make sure this is a component
+        match(trail-comp?):
+          (_:None): None()
+          (given:One<JITXObject>):
+            val c-or-m? = value(given)
+            if not is-component?(c-or-m?):
+              println("Component: %_" % [c-or-m?])
+              throw $ InstanceTypeError(expects = SingleComponent, observed = instance-type(c-or-m?))
+            given
+
+
+doc: \<DOC>
+Retrieve an Instance from a particular context by Ref
+
+This function must be called from within a `pcb-module`
+context.
+
+This function is not recursive. If you pass `r = some.obj`
```

Comment:
Looks like we're not using keyword args below. Perhaps either change the signature to use keyword args or rephrase the doc here to not reference passing in `r = ...`